Adds a is-empty/not-empty for strings

robdaemon opened the pull request:
What the UI calls "is empty" / "not empty" right now is a null check,
not a null and "" string check.

This adds is-empty / not-empty operators in the query processor, and
reworks the UI to use those. Any existing query that previously used the
is-null / not-null operators will show (as "Is null" / "Not null") but
any newly made queries will not be able to select "Is null" / "Not null"
for text type fields.

This will allow backwards compatibility with existing questions, while
promoting use of the new "is-empty"/"not-empty" operators.

Resolves #5299

Commit:
Adds a is-empty/not-empty for strings

What the UI calls "is empty" / "not empty" right now is a null check,
not a null and "" string check.

This adds is-empty / not-empty operators in the query processor, and
reworks the UI to use those. Any existing query that previously used the
is-null / not-null operators will show (as "Is null" / "Not null") but
any newly made queries will not be able to select "Is null" / "Not null"
for text type fields.

This will allow backwards compatibility with existing questions, while
promoting use of the new "is-empty"/"not-empty" operators.

Resolves #5299

paulrosenzweig commented on frontend/src/metabase/query_builder/components/filters/FilterPopoverHeader.jsx:
```diff
@@ -47,6 +47,7 @@ export default function FilterPopoverHeader({
           })}
+          field={field}
           operator={filter.operatorName()}
           operators={filter.filterOperators()}
```
I think it'd be better to pull the logic around text fields out of the UI. Two half-baked ideas for that:
1. Thread the current value down through `filterOperators()`, so that the code generating the list of operators for a given field type can make use of the current value.
2. Update this component to always display the current value in the list even if it isn't one of the `operators`.

robdaemon replied:
I'm having a little trouble with option 1 there - `filterOperators` ends up memoized, which seems to not like an `undefined` or empty key for an argument.

robdaemon replied:
wait, I figured it out :smile:

paulrosenzweig replied:
It looks like `memoize` expected a consistent number of arguments from the caller (ignoring any default arg values). I updated it to use the number of arguments as part of the cache key, which I think fixes the issue. I pushed that change on this branch.

Commit:
move operator filter from UI to metadata

This has some complications due to our memoization code - the default
value for the parameter must be something that can be used as a map key.
You cannot use null, {}, etc for that purpose. Used 0 as a magic value
here, but I'm open to other ideas.

paulrosenzweig committed:
update memoize util to allow variable number of args

Commit:
code review feedback

Moved filtering of null operators down to the schema_metadata level

Fixed flow warnings about annotations

Added Cypress test to make sure the proper operators show up when asking
questions

robdaemon commented:
turns out that `// $FlowFixMe` is still necessary:

```
frontend/src/metabase/query_builder/components/filters/FilterPopover.jsx:122
122:     this.setFilter(filter.set(newFilter));
                                   ^^^^^^^^^ Filter. This type is incompatible with the expected param type of
 15:   set(mbql: any[]) {
                 ^^^^^ array type. See: frontend/src/metabase-lib/lib/queries/structured/MBQLClause.js:15

frontend/src/metabase/query_builder/components/filters/FilterPopover.jsx:122
122:     this.setFilter(filter.set(newFilter));
                                   ^^^^^^^^^ null. This type is incompatible with the expected param type of
 15:   set(mbql: any[]) {
                 ^^^^^ array type. See: frontend/src/metabase-lib/lib/queries/structured/MBQLClause.js:15

frontend/src/metabase/query_builder/components/filters/FilterPopover.jsx:122
122:     this.setFilter(filter.set(newFilter));
                                   ^^^^^^^^^ undefined. This type is incompatible with the expected param type of
 15:   set(mbql: any[]) {
                 ^^^^^ array type. See: frontend/src/metabase-lib/lib/queries/structured/MBQLClause.js:15
```

sbelak reviewed (approved):
Looks good on the BE side

paulrosenzweig reviewed (approved):
LGTM!
